File: src/matlab/dft_split.m
function [xr, xi] = dft_split(s)
% Computation of the DFT of two real sequences (Splitting)
% Input:
%	s = DTF(x0 + i*x1)
% Output: 
%	xr = DFT(x0)
%	xi = DFT(x1)

N = numel(s);
xr = zeros(1, N);
xi = zeros(1, N);

xr(1) = real(s(1)) + 1i*0;
xi(1) = imag(s(1)) + 1i*0;
xr(N/2+1) = real(s(N/2+1));
xi(N/2+1) = imag(s(N/2+1));

for k=2:N/2
	xr(k) = (s(k)+conj(s(N-k+2)))/2;
	xi(k) = (s(k)-conj(s(N-k+2)))/2i;
end

for k=2:N/2
	xr(N/2+k) = conj(xr(N/2-k+2));
	xi(N/2+k) = conj(xi(N/2-k+2));
end

end

File: src/dft_split.c
/**
 * @file dft_split.c
 * @brief Computation of the DTF of two real sequences (Splitting)
 * @author matyunin.d
 * @date 28.08.2019
 */

#include <complex.h>

/**
 * @brief Split two real sequences
 * s(k) = DFT[ xr(k) + i*xi(k)] => DFT[xr(k)] and DFT[xi(k)]
 * @param s Pointer to complex-valued sequence with two real-valued sequences
 * @param xr Pointer to output 'xr' sequence
 * @param xy Pointer to output 'xy' sequence
 * @param N Length of the sequences
 */
void dft_split(const double complex *s, complex double *xr, complex double *xi, unsigned int N)
{
	xr[0] = creal(s[0]);
	xi[0] = cimag(s[0]);
	xr[N/2] = creal(s[N/2]);
	xi[N/2] = cimag(s[N/2]);

	for (unsigned int k = 1; k < N/2; k++) {
		xr[k] = (s[k] + conj(s[N-k])) / 2;
		xi[k] = (s[k] - conj(s[N-k])) / (2*I);
	}

	for (unsigned int k = 1; k < N/2; k++) {
		xr[N/2+k] = conj(xr[N/2-k]);
		xi[N/2+k] = conj(xi[N/2-k]);
	}
}


